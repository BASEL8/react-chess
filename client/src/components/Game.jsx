import React, { useEffect, useState } from 'react';
import Helmet from 'react-helmet';
import Chess from 'chess.js';
import Chessboard from 'chessboardjsx';
import axios from 'axios';
import { Link } from 'react-router-dom';

let apiUrl = 'http://localhost:8000/api';

function Game(props) {
  let [chess, updateChess] = useState(new Chess());
  let [fen, updateFen] = useState('start');
  let newFen = '';
  let [moveHistory, updateMoveHistory] = useState([]);
  let newHistory = [];
  let id = props.match.params.id;
  const [friends, setFriends] = useState([]);
  const [seeks, setSeeks] = useState([]);

  useEffect(() => {

    axios.get(`http://localhost:8000/api/game/${id}`)
    .then((response) => {
      // console.log("friends data: ",response.data);
      setFriends(response.data);
    });

    axios.get(`http://localhost:8000/api/seeks/${localStorage.getItem('userId')}`)
      .then((response) => {
        // console.log("Seek data: ", response.data);
        setSeeks(response.data);
      });
  }, []);

  //Check if there are two players
  //If there is a move history, add it to the Chess game
  useEffect(() => {
    axios.get(`${apiUrl}/game/${id}`)
    .then(({ data, err }) => {
      if (err) {
        return console.log(err);
      }
      console.log('Check for players and move history');

      if (data.playerOne, data.friendId) {
        if(data.history.length > 0) {
          console.log('history exist');
          let history = data.history;
          let historyWithNumbers = [];
          console.log('history: ', history);
          let firstMove = `1.${history[0]}`;
          let moveString = '';

          //Create new game
          const game = new Chess();

          //Converting move history into a string to use in pgn
          let nr = 1;
          let newString = '';
          for(let i = 0; i < history.length; i++) {
            if(i % 2 === 0) {
              newString = `${nr}.${history[i]}`;
              historyWithNumbers.push(newString);
              nr++;
            }
            else {
              historyWithNumbers.push(history[i]);
            }
          }
          moveString = historyWithNumbers.join(' ');

          let pgn = [
            moveString,
          ];

          //With 'white' and 'black' tags
          //let playerOne = data.playerOne;
          //let playerTwo = data.playerTwo;
          // pgn = [
          //   `[White "${playerOne}"]`,
          //   '[Black "Jean Dufresne"]',
          //   firstMove,
          // ];

          //Load moves to Chess + update states chess and fen
          game.load_pgn(pgn.join('\n'));
          updateChess(game);
          updateFen(game.fen());
          updateMoveHistory(game.history());
          //console.log(game);
        }
        else {
          console.log('no history');
        }
      }
    });
  }, []);

  //Check for fen updates every second
  useEffect(() => {
    const timer = setInterval(() => {
      //console.log('Checking for new fen.');
      axios.get(`${apiUrl}/game/${id}`)
      .then((response) => {
        newFen = response.data.fen;

        //If fen has changed, update chess, fen and history
        if(newFen && newFen !== fen) {

          //Adding move to chess
          newHistory = response.data.history;
          let latestMove = newHistory[newHistory.length - 1];
          chess.move(latestMove);

          //Updating fen and history
          updateFen(newFen);
          updateMoveHistory(newHistory);
        }
      });
    }, 1000);

    return () => {
      console.log('clean');
      return clearTimeout(timer);
    };
  });

  // do not pick up pieces if the game is over
  // only pick up pieces for the side to move
  // function onDragStart(source, piece, position, orientation) {
  //   if (chess.game_over() === true ||
  //       (chess.turn() === 'w' && piece.search(/^b/) !== -1) ||
  //       (chess.turn() === 'b' && piece.search(/^w/) !== -1)) {
  //     return false;
  //   }
  // };

  function onDrop(sourceSquare) {
    let sourceSq = sourceSquare.sourceSquare;
    let targetSq = sourceSquare.targetSquare;
    let piece = sourceSquare.piece;
    let pieceType = piece.split('')[1];
    let isCapture = chess.get(targetSq);
    let sloppyMove = '';

    if(isCapture) {
      sloppyMove = pieceType + sourceSq + 'x' + targetSq;
    }
    else {
      sloppyMove = pieceType + sourceSq + targetSq;
    }

    let move = chess.move(sloppyMove, {sloppy: true});

    //If the move is acceptable, send it to the server
    if(move) {

      //Creating variables
      newFen = chess.fen();
      newHistory = chess.history();

      //Update fen and history
      updateFen(newFen);
      updateMoveHistory(newHistory);

      //Send move to backend
      axios.post(`${apiUrl}/game/move`, {
        id: id,
        gameFen: newFen, //State fen cannot be used since it may not be updated yet
        gameHistory: newHistory,
        gameStyle: "standard",
      })
      .then(function (response) {
        //console.log('POST response in onDrop');
        //console.log(response);
      })
      .catch(function (error) {
        console.log(error);
      });
    }
    else {
      console.log('ILLEGAL MOVE');
    }
  }

  function onClickReset(e) {
    chess.reset();
    updateFen(chess.fen());

    //Send reset to backend
    axios.post(`${apiUrl}/game/move`, {
      id: id,
      gameFen: 'start',
      gameHistory: [],
      gameStyle: "standard",
    })
    .then(function (response) {
      //console.log(response);
    })
    .catch(function (error) {
      console.log(error);
    });
  }

  return (
    <div className="App">
      <Helmet><title>Game</title></Helmet>
      <Link to='/lobby' className="btn btn-primary"><button type="submit">Back to Lobby</button></Link>
      <p><b>PlayerOne: {friends.playerOne}</b> against <b>friendId: {friends.friendId}</b></p>
      <Chessboard
      position={fen}
      onDrop={onDrop}
      />
      <HistoryTable moveHistory={moveHistory}/>
      <button onClick={onClickReset}>Restart</button>
    </div>
  );
}

function HistoryTable(props) {
  let moveHistory = props.moveHistory;
  let whitePlayer = true;

  let tableRows = moveHistory.map( move => {
    if(whitePlayer) {
      return <tr><td>{move}</td></tr>;
    }
    else {
      return <td>{move}</td>;
    }
    whitePlayer = !whitePlayer;
  });

  return(
    <table>
      <thead>
        <tr>
          <th>Moves</th>
        </tr>
      </thead>
      <tbody>
        {tableRows}
      </tbody>
    </table>
  );
}

export default Game;
