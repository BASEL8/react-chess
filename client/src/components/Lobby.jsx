import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import ClearIcon from '@material-ui/icons/Clear';

const useStyles = makeStyles({
  wrapper: {
    diplay: 'flex',
    justifyContent: 'flex-start',
    flexWrap: 'nowrap'
  },
  table: {
    maxWidth: '60vw',
  },
  buttonContainer: {
    maxWidth: '25vw',
    display: 'flex',
    flexDirection: 'column',
    alignContent: 'center'
  }
});

function createData(id, player, fat, carbs, protein) {
  return { id, player, fat, carbs, protein };
}

const rows = [
  createData(1, 'Frozen yoghurt', 6.0, 24),
  createData(2, 'Ice cream sandwich', 9.0, 37),
  createData(3, 'Eclair', 16.0, 24),
  createData(4, 'Cupcake', 3.7, 67),
  createData(5, 'Gingerbread', 16.0, 49),
];

export default function Lobby () {
  const classes = useStyles();

  return (
    <div className={classes.wrapper}>
      <TableContainer component={Paper} className={classes.table}>
        <Table aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>No.</TableCell>
              <TableCell>Player</TableCell>
              <TableCell>Play</TableCell>
              <TableCell>Delete</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {rows.map(row => (
              <TableRow key={row.id}>
                <TableCell component="th" scope="row">
                  {row.id}
                </TableCell>
                <TableCell>{row.player}</TableCell>
                <TableCell><Button variant="outlined">Play</Button></TableCell>
                <TableCell><Button><ClearIcon /></Button></TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      <div className={classes.buttonContainer}>
        <Button variant="outlined">Create a new match</Button>
        <Button variant="outlined">Play with a friend</Button>
        <Button variant="outlined">Play against computer</Button>
      </div>
    </div>
  );
}
