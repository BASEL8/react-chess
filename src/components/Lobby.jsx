/* eslint-disable no-undef */
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import Helmet from 'react-helmet';

const Lobby = () => {
  const [seeks, setSeeks] = useState([]);

  useEffect(() => {
    axios.get(`http://localhost:8000/api/seeks/${localStorage.getItem('userId')}`)
      .then((response) => {
        // console.log(response.data);
        setSeeks(response.data);
      });
  }, []);

  if (!seeks) {
    return <p>Loading seeks...</p>;
  }

  return (
    <>
      <Helmet><title>Lobby</title></Helmet>
      <table>
        <thead>
          <tr>
            <th>Player One</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          {seeks.map((seek) => (
            <>
              <tr key={seek.id}>
                <td>{seek.playerOne}</td>
                <td>
                  <button type="submit">Play</button>
                  {' '}
                </td>
              </tr>
            </>
          ))}
        </tbody>
      </table>
    </>
  );
};

export default Lobby;
