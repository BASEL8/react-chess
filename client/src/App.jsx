/* import React, { useEffect } from 'react';
import axios from 'axios';
import './App.css';
import Lobby from './components/Lobby';

const App = () => {
  useEffect(() => {
    axios.get('http://localhost:8000/api/seeks').then((res) => {
      console.log(res);
    });
  }, []);
  useEffect(() => {
    axios.post('http://localhost:8000/api/seeks/').then((res) => {
      console.log(res);
    });
  }, []);
  useEffect(() => {
    axios.post('http://localhost:8000/api/seeks/333').then((res) => {
      console.log(res);
    });
  }, []);
  useEffect(() => {
    axios.get('http://localhost:8000/api/game/333').then((res) => {
      console.log(res);
    });
  }, []);
  useEffect(() => {
    axios.post('http://localhost:8000/api/game/333/2_3').then((res) => {
      console.log(res);
    });
  }, []);


  return (
    <div className="App">
      <header className="App-header">
        <h1>
          Let's play chess!
        </h1>
      </header>
      <Lobby />
    </div>
  );
}; */

import React from 'react';
import './App.css';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import Login from './components/Login';
import Lobby from './components/Lobby';

const App = () => (
  <div className="App">
    <Router basename={process.env.PUBLIC_URL}>
      <Route exact path="/" component={Login} />
      <Route path="/lobby" component={Lobby} />
    </Router>
  </div>
);

export default App;
