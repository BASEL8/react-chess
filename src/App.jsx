import React, { useEffect, useState } from 'react';
import Chess from 'chess.js';
import Chessboard from 'chessboardjsx';
import './App.css';
let chess = new Chess();
let pgn = chess.pgn();

function App() {
  let [fen, updateFen] = useState('start');
  let [moveHistory, updateMoveHistory] = useState([]);
  let newFen = '';

  function updateView() {
    newFen = chess.fen();
    updateFen(newFen);
  }

  function onDrop(sourceSquare) {
    let sourceSq = sourceSquare.sourceSquare;
    let targetSq = sourceSquare.targetSquare;
    let piece = sourceSquare.piece;
    let moveObject = {};
    let newFen = '';

    // console.log(sourceSq, targetSq, piece);

    moveObject = chess.move({from: sourceSq, to: targetSq});

    // console.log(moveObject);

    if(moveObject) {
      console.log('Legal move.');
      updateView();
      let updatedHistory = chess.history();
      updateMoveHistory(updatedHistory);
    }
    else {
      console.log('ILLEGAL MOVE');
    }
  }

  function onClickReset(e) {
    chess.reset();
    updateView();
  }

  return (
    <div className="App">
      <Chessboard
      position={fen}
      onDrop={onDrop}
      />
      <MovesTable moveHistory={moveHistory}/>
      <button onClick={onClickReset}>Restart</button>
    </div>
  );
}

function MovesTable(props) {
  let moveHistory = props.moveHistory;
  console.log(moveHistory);

  let tableRows = moveHistory.map( move => {
    return <tr><td>{move}</td></tr>;
  });

  return(
    <table>
      <thead>
        <tr>
          <th>Moves</th>
        </tr>
      </thead>
      <tbody>
        {tableRows}
      </tbody>
    </table>
  );
}

export default App;
