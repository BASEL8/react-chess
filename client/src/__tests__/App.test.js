import React from 'react';
import { mount, configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router'; 
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import App from '../App';
import Start from '../components/Start';
import Lobby from '../components/Lobby';
import Game from '../components/Game';

configure({adapter: new Adapter()});
render(<App />);

describe('App component', () => {
  it('renders Start component on root', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/' ]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.find(Start)).toHaveLength(1);
  });

  it('routes /lobby to Lobby', () => {
    const component = shallow(<App />);
    expect(component.find('Route[path="/lobby"]').first().prop('component')).toBe(Lobby);
  });

  it('routes /game to Game', () => {
    const component = shallow(<App />);
    expect(component.find('Route[path="/game"]').first().prop('component')).toBe(Game);
  });

  /* it('renders Lobby component', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/lobby' ]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.find(Lobby)).toHaveLength(1);
  });

  it('renders Lobby component', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/game' ]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.find(Game)).toHaveLength(1);
  }); */
});
