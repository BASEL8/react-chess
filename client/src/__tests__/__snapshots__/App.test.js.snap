// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot renders 1`] = `undefined`;
